### ISSUE TITLE
Ingress not working with multi node setup

### ISSUE THREAD

**Issue by AlbertoDePena** [canonical/microk8s#1049]
I am running multi node as seen below... For some reason, I cannot get the ingress working. If I disable ingress addon and use NodePort service on an application replicated 3 times, I can get to the app by using k8s-worker-0X (where X is 1,2,3) with no problem. I dont have any dns entries in my /etc/hosts file and ip6 is enabled (I tried disabling it and couldnt enable metrics-server).

Is there some configuration in terms of dns entries or networking to get ingress (127.0.0.1) working? Thanks in advance.

Maybe some guidance on how to configure Ubuntu VMs to have microk8s work with multi node setup....

NAME            STATUS   ROLES    AGE   VERSION
k8s-master-01   Ready    <none>   31d   v1.17.3
k8s-worker-01   Ready    <none>   31d   v1.17.3
k8s-worker-02   Ready    <none>   31d   v1.17.3
k8s-worker-03   Ready    <none>   31d   v1.17.3

Inspecting services
  Service snap.microk8s.daemon-cluster-agent is running
  Service snap.microk8s.daemon-flanneld is running
  Service snap.microk8s.daemon-containerd is running
  Service snap.microk8s.daemon-apiserver is running
  Service snap.microk8s.daemon-apiserver-kicker is running
  Service snap.microk8s.daemon-proxy is running
  Service snap.microk8s.daemon-kubelet is running
  Service snap.microk8s.daemon-scheduler is running
  Service snap.microk8s.daemon-controller-manager is running
  Service snap.microk8s.daemon-etcd is running
  Copy service arguments to the final report tarball
Inspecting AppArmor configuration
Gathering system information
  Copy processes list to the final report tarball
  Copy snap list to the final report tarball
  Copy VM name (or none) to the final report tarball
  Copy disk usage information to the final report tarball
  Copy memory usage information to the final report tarball
  Copy server uptime to the final report tarball
  Copy current linux distribution to the final report tarball
  Copy openSSL information to the final report tarball
  Copy network configuration to the final report tarball
Inspecting kubernetes cluster
  Inspect kubernetes cluster

Building the report tarball
  Report tarball is at /var/snap/microk8s/1254/inspection-report-20200323_213735.tar.gz

microk8s is running
addons:
cilium: disabled
dashboard: enabled
dns: enabled
fluentd: disabled
gpu: disabled
helm: enabled
ingress: disabled
istio: disabled
jaeger: disabled
juju: disabled
knative: disabled
kubeflow: disabled
linkerd: disabled
metallb: disabled
metrics-server: enabled
prometheus: disabled
rbac: disabled
registry: disabled
storage: disabled

**Comment by balchua** [comment]
a few questions here, 
when you enable ingress, do you see the ingress controller running in the `ingress` namespace?
When you try to access your application, are you accessing your application from a remote machine? 
Thanks

**Comment by AlbertoDePena** [comment]
Yes to both. I'm sure the issue is me accessing the ingress from another machine. What's weird is that I can access it using one of the worker nodes (even thought connection drops after) I can't never access it from the master node.

**Comment by balchua** [comment]
Can help give me an example on how it works on a worker node? Like using `curl`?

Normally how i access ingress is via virtual host, say i define `myapp.example.local` into my remote machine's `/etc/host`.

**Comment by AlbertoDePena** [comment]
So I can (some times) curl or on a browser do this: http://k8s-worker-01/my-app/ (see nodes above) but k8s-master-01 always yields not found error.

**Comment by balchua** [comment]
Mind posting the ingress definition? Thanks

**Comment by AlbertoDePena** [comment]
```yaml
kind: Ingress
apiVersion: extensions/v1beta1
metadata:
  name: localhost-ingress
  annotations:    
    nginx.ingress.kubernetes.io/ssl-redirect: "false"
    nginx.ingress.kubernetes.io/force-ssl-redirect: "false"
    nginx.ingress.kubernetes.io/rewrite-target: /$2
spec:
  rules:
  - http:
      paths:
      - path: /azure-func-app-demo(/|$)(.*)
        backend:
          serviceName: azure-func-app-demo-service
          servicePort: 80
```

Then I try `curl http://k8s-master-01/azure-func-app-demo/`, always get `404`. Sometime `curl http://k8s-worker-01/02/03/azure-func-app-demo/` works.

Note that I curl within the `k8s-master-01` ubuntu VM and try navigating using a browser from outside the ubuntu VM. Same behavior

**Comment by balchua** [comment]
Can we try something simple?

```yaml
kind: Ingress
apiVersion: extensions/v1beta1
metadata:
  name: localhost-ingress
  annotations:    
    nginx.ingress.kubernetes.io/ssl-redirect: "false"
    nginx.ingress.kubernetes.io/force-ssl-redirect: "false"
    nginx.ingress.kubernetes.io/use-regex: "true"
spec:
  rules:
  - http:
      paths:
      - path: /azure-func-app-demo(/|$)(.*)
        backend:
          serviceName: azure-func-app-demo-service
          servicePort: 80
```

**Comment by AlbertoDePena** [comment]
I ll give that a shot tomorrow morning. Thank you @balchua  so much for your quick response =)

Anything special needed in terms of `/etc/hosts` files that I should know about?

**Comment by balchua** [comment]
Didn't do anything special with my kubernetes nodes' `/etc/host`.

**Comment by AlbertoDePena** [comment]
@balchua trying to ingress definition you provide I know get this error from any of the VMs

Multi-Hop Cycle Detected
Description: Your request is prohibited because it would cause a cycle.

**Comment by balchua** [comment]
You get this error when using curl?
Can we even make the ingress simpler by removing the regex? And maybe use `host` as part of the ingress rule?

**Comment by AlbertoDePena** [comment]
THis is so weird... so I  have master node `k8s-master-01` and 3 workers `k8s-worker-01/02/03` I have the same app replicated in `k8s-worker-01/02/03`. Based on the ingress above:

This never works:

`http://k8s-master-01/ro-extraction-queuer/' (404 Not Found)

This works sometimes, most of the time I get 404 Not Found

`http://k8s-worker-01/ro-extraction-queuer/`
`http://k8s-worker-02/ro-extraction-queuer/`
`http://k8s-worker-03/ro-extraction-queuer/`

Are we sure the ingress addon works with multi node cluster?

**Comment by balchua** [comment]
Yes ingress on multi node works for me.  I've tested it many times.

If it is ok, can you attach the `inspect` tarball?

**Comment by AlbertoDePena** [comment]
It has to be networking issues. Do you have DNS entries for each node in your /etc/hosts files? What about ip6?

**Comment by balchua** [comment]
I disabled ipv6.  So in order to use the ingress ` host` rule, i have to add my fake uri in `/etc/hosts` and thats about it.
Fyi, i have always used the `host` rule though.

**Comment by AlbertoDePena** [comment]
So I thought if we disable ip6 other add-ons won't work like metrics server. Also by host in etc/hosts you mean adding the fake host mapping to your master node IP?

**Comment by balchua** [comment]
I didn't try metrics server yet on multi node.  And yes the fake host points to any of the worker node running ingress.

**Comment by AlbertoDePena** [comment]
So I am getting closer... now I can get it working although every once in a while it returns 404. Since I have a 4 node cluster, ingress addon deployed a daemonset which created 4 ingress pods (one per node). Is this expected or should there only be on ingress pod in the master node?

**Comment by ktsakalozos** [comment]
@AlbertoDePena ingress is a Daemonset [1]. You should have one ingress pod running on each node.

[1] https://github.com/ubuntu/microk8s/blob/master/microk8s-resources/actions/ingress.yaml#L125

**Comment by pastarace** [comment]
I'm in the same situation of @AlbertoDePena . I've a 3 nodes cluster with the ingress enabled. My deployments, services and ingresses are correctly setup and running in a dedicated namespace.

While my expectation was to reach my services always using the master node ip having the ingress acting as a router/loadbalancer, I noticed that since the ingress pods are running on each node (as @ktsakalozos correctly pointed out), I can reach the service with a single pod deployed on node X only targeting the ip of node X (using of course the hostname I defined and not the ip itself, having it mapped in the /etc/hosts file).

Comparing with the official [ingress-nginx deployment guide for bare-metal](https://kubernetes.github.io/ingress-nginx/deploy/#bare-metal), it defines an ingress-nginx-controller NodePort service, while microk8s version doesn't. Do I need to enable it to always be able to use a single ip to access my services externally?

Thanks

**Comment by pastarace** [comment]
Moreover they use a Deployment (with one replica) instead of a DeamonSet, having a single point o ingress into the cluster

**Comment by pastarace** [comment]
I solved creating a service of type NodePort to be able to access the ingress controllers from any node of the cluster.

As described [here](https://www.net7.be/blog/article/kubernetes_ingress_bare_metal_load_balancing.html), I should then use in front a loadbalancer like HAproxy to balance the usage of the service on all the nodes.

Could it be a reference architecture for the ingress use case in microk8s?

Thanks

```
apiVersion: v1
kind: Service
metadata:
  name: ingress-nginx
  namespace: ingress
  labels:
    name: nginx-ingress-microk8s
spec:
  type: NodePort
  ports:
    - name: http
      port: 80
      targetPort: 80
      nodePort: 31080
      protocol: TCP
    - name: https
      port: 443
      targetPort: 443
      nodePort: 31443
      protocol: TCP
  selector:
    name: nginx-ingress-microk8s
```

**Comment by zazizou** [comment]
Hello Guys,

I am in a similar situation. Though I was trying to set up a multi-node ingress load balanced with Metallb. I have enabled ingress addon then changed the hostPort from 80 to 8880 (http) and from 443 to 4443 (https) because I need 80 and 443 ports. I have then enabled metallb and expose a service like #https://github.com/ubuntu/microk8s/issues/1049#issuecomment-682592560 by @pastarace but of type LoadBalancer.

I am using 5 nodes cluster in HA  :  mac0 and mac[1-4] which are joined to mac0 

Here is what I discover:

-  On a single node mac0 (before joining other nodes) everything work as expected: path routed ingress and host routed ingress just work and I can spawn other services as type LoadBalancer as needed.
- On multi node: host routed ingress only works on pods that are not deployed on mac0 (master) and path routed ingress works only on the node where the pod is deployed. I have checked that I can expose any deployment as a LoadBalancer type, and it works independently of nginx ingress.

I have started from scratch several times now, trying to understand what is happening that I found myself lost and without any other options.

Multi node ingress is a mess
